sdks/python/apache_beam/runners/common.py

```diff
@@ -323,6 +330,27 @@ def is_splittable_dofn(self):
     # type: () -> bool
     return self.get_restriction_provider() is not None
 
+  def is_tracking_watermark(self):
+    return self.get_watermark_estimator_provider() is not None
+
+  def get_restriction_coder(self):
+    """Get coder for a restriction when processing an SDF.
+
+    If the SDF uses an WatermarkEstimatorProvider, the restriction coder is a
```

Comment:
I wonder if it would be preferable to have a no-op estimator if one is not provided, rather than have if statements here (and elsewhere) to handle two separate cases.

Reply:
On this note, are we correctly producing tuples only when we should? (Or is the restriction coder by default pickle_coder, in which case we might not notice by default?)

Reply:
Within current changes, we only produce the tuple when we should. E.g., if there is a watermark_estimator, return (element, (restriction, estimator_state)). If there is no watermark_estimator provided, return (element, restriction). So does the restriction_coder.
Providing a no-op estimator is a good idea. I'm wondering whether it's also feasible for other SDK, like using `None` as estimator_state.